packages/service/lib/dependencies.js

```diff
@@ -0,0 +1,116 @@
+'use strict'
+const { readFile } = require('node:fs/promises')
+const { resolve } = require('node:path')
+const { closest } = require('fastest-levenshtein')
+const Topo = require('@hapi/topo')
+const ConfigManager = require('@platformatic/config')
+const errors = require('./errors')
+
+function missingDependencyErrorMessage (clientName, service, services) {
+  const closestName = closest(clientName, [...services.keys()])
+  let errorMsg = `service '${service.id}' has unknown dependency: '${clientName}'.`
+  if (closestName) {
+    errorMsg += ` Did you mean '${closestName}'?`
+  }
+  return errorMsg
+}
+
+async function getClientId (service, client) {
+  try {
+    const clientPackageJsonPath = resolve(service.path, client.path, 'package.json')
+    const clientPackageJSON = JSON.parse(await readFile(clientPackageJsonPath, 'utf8'))
+    return clientPackageJSON.name ?? ''
+  } catch (err) {
+    if (client.url === undefined || client.name === undefined) {
+      throw err
+    }
+  }
+}
+
+async function parseConfigUrl (configManager, urlString) {
+  if (!urlString) {
+    return { url: null, envVar: null }
+  }
+
+  try {
+    const url = await configManager.replaceEnv(urlString)
+    return { url, envVar: null }
+  } catch (err) {
+    // The MissingValueError is an error coming from pupa
+    // https://github.com/sindresorhus/pupa#missingvalueerror
+    // All other errors are simply re-thrown.
+    if (err.name !== 'MissingValueError' || urlString !== `{${err.key}}`) {
+      throw err
+    }
+
+    return { url: null, envVar: err.key }
+  }
+}
+
+async function addDependency (configManager, services, service, dependencyId, urlString) {
+  let { url, envVar } = await parseConfigUrl(configManager, urlString)
+  if (url !== null) {
+    service.dependencies.push({ id: dependencyId, url, local: false })
+    return
+  }
+
+  const dependency = services.get(dependencyId)
+  if (typeof dependency === 'undefined') {
+    throw new errors.MissingDependencyError(
+      missingDependencyErrorMessage(dependencyId, service, services)
+    )
+  }
+
+  url = `http://${dependencyId}.plt.local`
+
+  if (envVar !== null) {
+    service.localServiceEnvVars.set(envVar, url)
+  }
+
+  dependency.dependents.push(service.id)
+  service.dependencies.push({ id: dependencyId, url, local: true })
+}
+
+async function getDependencies (services, service) {
```

Comment:
IMHO this function should not know anything about other services. It should accept a service config and an `onMissingEnv` callback function and return a list of dependencies some format like `[{ url }]`.

Then runtime should pass a `onMissingEnv` where it would inject an env var if it's missing. When the runtime receives the list of parsed dependencies it can define it this deplendency local or not and what is the dependency id.

Reply:
In the original implementation `services` was used first of to check if the target dependency exists, and then to add the current service as dependent. If we give up on tracking the dependent one I can probably simplify services to be a list of ids.
But I think a service should be aware of other services.

Reply:
Runtime can put the service in the dependent list after it receives dependencies from all services, by matching internal services domain. But I'm +1 of get rid of the dependent property.

[Here](https://github.com/platformatic/platformatic/pull/2665#issue-2338467821) was my last attempt of dropping it.

Reply:
Ok, I'll see I can remove the services in the functions.

Reply:
@ivan-tymoshenko @mcollina  I reworked everything. Is this good to go now?